### ISSUE TITLE
Opaque vs. Qed

### ISSUE THREAD

**Issue by benediktahrens** [UniMath/UniMath#548]
It has been advertised to never close proofs with `Qed`, and to opacify with `Opaque` afterwards instead.
It has also been pointed out that those two options are not equivalent. With @amblafont , we have now stumbled over such a situation, where a constant opacified afterwards with `Opaque` makes Coq loop, whereas a `Qed`'d constant of the same type works.

@amblafont has prepared an example file, which you can get from [1].

Any comments are appreciated.

[1] https://raw.githubusercontent.com/amblafont/largecatmodules/33f56b3fe7f3ac70225a97aeb8fcace7cfba0af5/Modules/Prelims/pbqed.v

**Comment by JasonGross** [comment]
Note that the kennel basically ignores `Opaque`, as do tactics like `change` and `unify`.  It basically means "don't unfold me in `cbv`, `lazy`, `simpl`, `cbn`, `unfold`".  By contrast only things like `Print`, `Print Assumptions`, and `Extraction` will bypass `Qed`.

**Comment by mortberg** [comment]
Because of this I think the UniMath style guide should be changed to say that `Qed` should be used whenever it is suitable for efficiency reasons. If Coq would have some kind of way of locking and unlocking definitions that would be great, but `Opaque`/`Transparent` is not really doing that. 

Note that there are other ways of locking definitions than `Qed` which is used heavily in SSReflect/MathComp, see the comment below this Definition:

http://ssr.msr-inria.inria.fr/doc/ssreflect-1.5/Ssreflect.ssreflect.html#ssr_converse

It might be worth understanding this and implementing it in UniMath in order to improve compilation times. The drawback is that it will obfuscate the code, but if UniMath is going to continue grow this might be necessary unless Coq changes.

**Comment by DanGrayson** [comment]
Re: "`Qed` should be used whenever it is suitable for efficiency reasons"

Well, no, because future users may want expressions involving your code to compute.  I would
prefer using `Qed` as seldom as possible.

**Comment by DanGrayson** [comment]
Benedikt, I guess you didn't literally mean "loop", but rather, "take a long time".

**Comment by benediktahrens** [comment]
On 01/20/2017 03:07 PM, Daniel R. Grayson wrote:
> Benedikt, I guess you didn't literally mean "loop", but rather, "take
> a long time".

The honest answer is that I don't know what Coq is doing, but yes, it
takes a long time. We have not waited long enough to see if it terminates.

**Comment by DanGrayson** [comment]
It might be illuminating to run Coq in the debugger and see what it's doing.

**Comment by benediktahrens** [comment]
On 01/20/2017 04:11 PM, Daniel R. Grayson wrote:
> It might be illuminating to run Coq in the debugger and see what it's
> doing.

Well, we know that it is spending a lot of time on examining a proof of
`isaset`, which we don't want it to do.

I am not sure I'd find the details of that activity to be very illuminating.

**Comment by DanGrayson** [comment]
Yes, but why does it examine it first, before doing what it would have done if the proof
hadn't been opaque?  That might be interesting.

**Comment by benediktahrens** [comment]
On 01/20/2017 04:37 PM, Daniel R. Grayson wrote:
> Yes, but why does it examine it first, before doing what it would have done if the proof
> hadn't been opaque?  That might be interesting.

That has to do with the evaluation strategy, I think. Eager vs lazy, or
something similar.

**Comment by DanGrayson** [comment]
This is such a simple concrete example that we might be able to recommend an
improvement in the evaluation strategy if we understood what it's doing.

**Comment by DanGrayson** [comment]
I've assigned myself to this issue, so I can be reminded to look into that in the debugger.

**Comment by benediktahrens** [comment]
On 01/20/2017 04:43 PM, Daniel R. Grayson wrote:
> I've assigned myself to this issue, so I can be reminded to look into
> that in the debugger.

That's great, thanks a lot!

**Comment by vladimirias** [comment]
I am with Dan on this question.

At the same time I think we should start prodding Matthew Sozeau to introduce a “transparent” for Qed.

There is one additional reason to prefer Opaque. When Coq takes a long time to check something there is usually a way to re-arange the code to improve the compilation time and this re-arrangement besides improving the compilation time also makes the code structurally more elegant.

**Comment by DanGrayson** [comment]
What does rearranging the code to improve the compilation time have to do with `Opaque`?

**Comment by vladimirias** [comment]
The argument against opaque is that it does not solve the problem of slow code. I commented that this may be a signal that the code can be re-arranged to make it more readable and at the same time faster.

Many a time I was trying to resolve the slowness issue and as a result found a more elegant solution for the code.

**Comment by JasonGross** [comment]
I would hypothesize that when something is slow and would be sped up by making something `Qed`, this suggests that what you're proving is an instance of a more general lemma, which does not have any access to the details of the opaque thing you're using.  Vladimir, does this match with your experience?

**Comment by vladimirias** [comment]
> which does not have any access

You mean does not need to have any access?

**Comment by JasonGross** [comment]
I mean both.

**Comment by mortberg** [comment]
One argument for using `Qed` a lot is that tracking down one `Qed` too many is easier than tracking down one `Qed` too little. If there is something that doesn't reduce because of a `Qed` I can partially evaluate my term and usually find it, if applying a tactic or a `Defined` takes very long it is often hard to figure out why.

**Comment by benediktahrens** [comment]
Let's discuss scientific evidence instead of religious principles:

Here is where Qed'ing has helped in the past:

- The inductive sets Anders and I have programmed do compute as well as one can reasonably hope for. They are heavily based on stuff in CT that uses Qed a lot.
Anders will confirm that Qed'ing was crucial.
- The example mentioned in the initial post of this issue is one where Qed makes things work, whereas Defined + Opaque does not.
- With Ralph, when programming the core of SubstitutionSystems, we had trouble getting code to compile in reasonable time, before using Qed everywhere we could.


Now, in each single case, you might argue that one should have changed the formalization or the maths, and things would have worked out fine with everything Defined. But that argument is worthless without any evidence.


On the other hand, I have not yet seen an example of something not computing due to a proof of isaset being Qed'd. Can you point me to such an example?

**Comment by DanGrayson** [comment]
I just did an experiment: I made `isasetsetquot` Qed'd, and everything seems to still work.  

I suppose this is for one of two reasons: we never have to prove that two hSets coming from setquot are equal; or all our code makes hSets from setquot using `isasetsetquot`, as in `setquotinset`.  The latter suffices.

By contrast, two elements of `std 3` may easily have proofs of `i<3` that are constructed differently, and so should not be Qed'd inside them, as we often have to prove two such elements are equal, and reflexivity is more convenient than using that `i<3` is a proposition.

**Comment by mortberg** [comment]
Let me elaborate a bit on what I meant by using `Qed` "whenever suitable". In my experience it is often fine to close proofs of propositions with `Qed`, except for a few exceptions like `iscontr` where one needs to be able to extract the center of contraction for things to compute properly. In CT I always close proofs of equalities between morphisms with `Qed` as the hom-sets are sets, I never had any troubles because of this. So if one is careful one can often have large parts of terms opaque and not have any trouble with computation. The benefit of this is that unification (which is used when in many tactics (apply, rewrite...), to infer implicit arguments, etc.) and conversion (equality checking during typechecking) is much faster as Coq can't unfold these propositions.

**Comment by mortberg** [comment]
I discussed this with Enrico Tassi (one of the Coq developers) today. He told me that it would be easy to implement a version of Qed that can be made transparent (but not opaque again). This way we could have things Qed'd and then make them transparent later if something needs the body of the term for computation. Would this something we want for UniMath? If so he can maybe make a branch of Coq for us with this feature so that we can experiment with it.

**Comment by vladimirias** [comment]
We made a lot of effort to make UniMath run on the of-the-shelf Coq and we certainly don’t want to get back to the state where we are on a branch or where patches are needed unless this gives us a very important advantage.

Having a Qed that can be “opened up” *but not closed up again” does not sound very useful.

**Comment by benediktahrens** [comment]
Re "Having a Qed that can be “opened up” *but not closed up again” does not sound very useful." 
Closing up a term `t` can make that another term that was previously well-typed (due to some computation involving `t` happening) becomes ill-typed. That would be weird, and I guess that that is the reason for this restriction.

**Comment by mortberg** [comment]
Yes, exactly. I thought I had answered with exactly this, but apparently I was mistaken. I guess a fancy way of saying this is that it breaks subject reduction. However going from something opaque to transparent is safe.

**Comment by peterlefanulumsdaine** [comment]
Returning to this old issue: as far as I can see, no-one was arguing in favour of the current situation, where `Qed.` is prohibited but `Opaque.` and `Abstract.` are allowed.

So, here is a suggestion: replace the current blanket prohibition 

> Do not end a proof with `Qed.`, except with `Goal`, for that may prevent later computations.

with a recommendation

> Generally, avoid creating opaque proofs (i.e. `Qed`, `Opaque`, and `abstract`), since these block computation.  Use these only when you have a strong positive reason (e.g. compilation speed), and are confident that computation will rarely/never be needed.

If the maintainers agree, I will make a PR of something like this.

**Comment by peterlefanulumsdaine** [comment]
@mortberg : I would suggest not including those.  The guidelines in the README are currently both pretty simple and pretty non-controversial, and it seems good to try to stick to that.  These suggestions seem a bit over-detailed, and also a bit beyond the consensus recommendation of this thread — e.g. as @DanGrayson argued upthread, it’s very often helpful to leave equalities transparent (in e.g. `is_nat_trans`), since they may turn out to be reflexivity, allowing transports to compute down the line.

**Comment by mortberg** [comment]
After discussing this with Peter in person I agree with what he says and that we shouldn't include this kind of technical details in the README. So I'm in favour of doing he changes Peter suggests.

**Comment by benediktahrens** [comment]
On 12/13/2017 03:56 PM, Peter LeFanu Lumsdaine wrote:
> Returning to this old issue: as far as I can see, no-one was arguing in favour of the current situation, where `Qed.` is prohibited but `Opaque.` and `Abstract.` are allowed.
> 
> So, here is a suggestion: replace the current blanket prohibition
> 
>> Do not end a proof with `Qed.`, except with `Goal`, for that may prevent later computations.
> 
> with a recommendation
> 
>> Generally, avoid creating opaque proofs (i.e. `Qed`, `Opaque`, and `abstract`), since these block computation.  Use these only when you have a strong positive reason (e.g. compilation speed), and are confident that computation will rarely/never be needed.
> 
> If the maintainers agree, I will make a PR of something like this.
> 


I use Qed myself all the time - I just suggested using "abstract" since 
it is less intrusive here, maybe.

**Comment by DanGrayson** [comment]
Thanks for letting me know.  I'll mark that down as an option.

Yes, UniMath is still using type-in-type, but I'd like to get polymorphic universes working.  Still haven't heard from anyone about https://github.com/coq/coq/issues/8196
